File: src/domain/models/ticket_values.py
"""Value objects for the Ticket Pipeline bounded context.

TicketDetailLevel maps SubdomainClassification to generation depth.
Tier and TierClassification implement two-tier generation: near-term
tickets (depth ≤2) keep classification-based detail, far-term (depth >2)
are downgraded to STUB. Core subdomain overrides depth (always FULL).

GeneratedEpic, GeneratedTicket, and DependencyOrder are frozen dataclasses
representing ticket pipeline artifacts.
"""

from __future__ import annotations

import enum
from dataclasses import dataclass
from typing import TYPE_CHECKING

if TYPE_CHECKING:
    from src.domain.models.domain_values import SubdomainClassification


class TicketDetailLevel(enum.Enum):
    """Generation depth for a ticket, driven by subdomain classification.

    FULL (Core):       All sections -- Goal, DDD Alignment, Design, SOLID Mapping,
                       TDD Workflow, Steps, AC, Edge Cases, Quality Gates.
    STANDARD (Supporting): Core sections -- Goal, DDD Alignment, Steps, AC, Quality Gates.
    STUB (Generic):    Minimal -- Goal + single AC.
    """

    FULL = "full"
    STANDARD = "standard"
    STUB = "stub"

    @staticmethod
    def from_classification(classification: SubdomainClassification) -> TicketDetailLevel:
        """Map a SubdomainClassification to its TicketDetailLevel."""
        from src.domain.models.domain_values import SubdomainClassification

        mapping = {
            SubdomainClassification.CORE: TicketDetailLevel.FULL,
            SubdomainClassification.SUPPORTING: TicketDetailLevel.STANDARD,
            SubdomainClassification.GENERIC: TicketDetailLevel.STUB,
        }
        return mapping[classification]


# ---------------------------------------------------------------------------
# Two-tier generation (alty-2j7.11)
# ---------------------------------------------------------------------------

_NEAR_TERM_MAX_DEPTH = 2


class Tier(enum.Enum):
    """Near-term vs far-term classification for two-tier generation."""

    NEAR_TERM = "near_term"
    FAR_TERM = "far_term"


@dataclass(frozen=True)
class TierClassification:
    """Depth-based tier with human-readable reason.

    Attributes:
        tier: Whether this ticket is near-term or far-term.
        reason: Why this classification was assigned.
    """

    tier: Tier
    reason: str


def classify_tier(
    depth: int,
    classification: SubdomainClassification,
) -> TierClassification:
    """Classify a ticket as near-term or far-term based on depth and subdomain.

    Rules:
        1. Core subdomain → always near-term (regardless of depth).
        2. depth ≤ 2 → near-term.
        3. depth > 2 → far-term.

    Args:
        depth: Hop count from dependency root (0 = no dependencies).
        classification: Subdomain classification of the ticket's bounded context.

    Returns:
        TierClassification with tier and reason.
    """
    from src.domain.models.domain_values import SubdomainClassification

    if classification == SubdomainClassification.CORE:
        return TierClassification(Tier.NEAR_TERM, "Core subdomain: always near-term")
    if depth <= _NEAR_TERM_MAX_DEPTH:
        return TierClassification(Tier.NEAR_TERM, f"Depth {depth} <= {_NEAR_TERM_MAX_DEPTH}")
    return TierClassification(Tier.FAR_TERM, f"Depth {depth} > {_NEAR_TERM_MAX_DEPTH}")


def tier_to_detail_level(
    tier: TierClassification,
    classification: SubdomainClassification,
) -> TicketDetailLevel:
    """Map a TierClassification to a TicketDetailLevel.

    Core subdomain is always FULL (invariant: Core tickets must have full
    AC, TDD phases, and SOLID mapping). Far-term non-Core tickets are
    always STUB. Near-term tickets use classification-based mapping.

    Args:
        tier: The tier classification.
        classification: Subdomain classification of the bounded context.

    Returns:
        The appropriate TicketDetailLevel.
    """
    from src.domain.models.domain_values import SubdomainClassification

    if classification == SubdomainClassification.CORE:
        return TicketDetailLevel.FULL
    if tier.tier == Tier.FAR_TERM:
        return TicketDetailLevel.STUB
    return TicketDetailLevel.from_classification(classification)


@dataclass(frozen=True)
class GeneratedEpic:
    """An epic grouping tickets for one bounded context.

    Attributes:
        epic_id: Unique identifier for this epic.
        title: Human-readable epic title.
        description: Epic description summarizing the bounded context scope.
        bounded_context_name: Which bounded context this epic covers.
        classification: Subdomain classification driving detail level.
    """

    epic_id: str
    title: str
    description: str
    bounded_context_name: str
    classification: SubdomainClassification


@dataclass(frozen=True)
class GeneratedTicket:
    """A generated beads ticket for an aggregate within a bounded context.

    Attributes:
        ticket_id: Unique identifier for this ticket.
        title: Human-readable ticket title.
        description: Rendered ticket body (detail depends on level).
        detail_level: Generation depth (FULL/STANDARD/STUB).
        epic_id: Parent epic this ticket belongs to.
        bounded_context_name: Which bounded context this ticket targets.
        aggregate_name: Which aggregate this ticket implements.
        dependencies: IDs of tickets that must be completed first.
    """

    ticket_id: str
    title: str
    description: str
    detail_level: TicketDetailLevel
    epic_id: str
    bounded_context_name: str
    aggregate_name: str
    dependencies: tuple[str, ...] = ()
    depth: int = 0


@dataclass(frozen=True)
class DependencyOrder:
    """Topologically sorted ticket execution order.

    Attributes:
        ordered_ids: Ticket IDs in dependency-safe execution order
                     (foundation first, dependents later).
    """

    ordered_ids: tuple[str, ...]

File: internal/ticket/domain/ticket_values.go
// Package domain implements the Ticket Pipeline bounded context domain layer.
package domain

import (
	vo "github.com/alty-cli/alty/internal/shared/domain/valueobjects"
)

// GeneratedEpic is an epic grouping tickets for one bounded context.
type GeneratedEpic struct {
	epicID             string
	title              string
	description        string
	boundedContextName string
	classification     vo.SubdomainClassification
}

// NewGeneratedEpic creates a GeneratedEpic value object.
func NewGeneratedEpic(
	epicID, title, description, boundedContextName string,
	classification vo.SubdomainClassification,
) GeneratedEpic {
	return GeneratedEpic{
		epicID:             epicID,
		title:              title,
		description:        description,
		boundedContextName: boundedContextName,
		classification:     classification,
	}
}

// EpicID returns the epic identifier.
func (e GeneratedEpic) EpicID() string { return e.epicID }

// Title returns the epic title.
func (e GeneratedEpic) Title() string { return e.title }

// Description returns the epic description.
func (e GeneratedEpic) Description() string { return e.description }

// BoundedContextName returns which bounded context this epic covers.
func (e GeneratedEpic) BoundedContextName() string { return e.boundedContextName }

// Classification returns the subdomain classification.
func (e GeneratedEpic) Classification() vo.SubdomainClassification { return e.classification }

// GeneratedTicket is a generated beads ticket for an aggregate.
type GeneratedTicket struct {
	ticketID           string
	title              string
	description        string
	detailLevel        vo.TicketDetailLevel
	epicID             string
	boundedContextName string
	aggregateName      string
	dependencies       []string
	depth              int
}

// NewGeneratedTicket creates a GeneratedTicket value object.
func NewGeneratedTicket(
	ticketID, title, description string,
	detailLevel vo.TicketDetailLevel,
	epicID, boundedContextName, aggregateName string,
	dependencies []string,
	depth int,
) GeneratedTicket {
	deps := make([]string, len(dependencies))
	copy(deps, dependencies)
	return GeneratedTicket{
		ticketID:           ticketID,
		title:              title,
		description:        description,
		detailLevel:        detailLevel,
		epicID:             epicID,
		boundedContextName: boundedContextName,
		aggregateName:      aggregateName,
		dependencies:       deps,
		depth:              depth,
	}
}

// TicketID returns the ticket identifier.
func (t GeneratedTicket) TicketID() string { return t.ticketID }

// Title returns the ticket title.
func (t GeneratedTicket) Title() string { return t.title }

// Description returns the rendered ticket body.
func (t GeneratedTicket) Description() string { return t.description }

// DetailLevel returns the generation depth.
func (t GeneratedTicket) DetailLevel() vo.TicketDetailLevel { return t.detailLevel }

// EpicID returns the parent epic ID.
func (t GeneratedTicket) EpicID() string { return t.epicID }

// BoundedContextName returns which bounded context this ticket targets.
func (t GeneratedTicket) BoundedContextName() string { return t.boundedContextName }

// AggregateName returns which aggregate this ticket implements.
func (t GeneratedTicket) AggregateName() string { return t.aggregateName }

// Depth returns the dependency depth.
func (t GeneratedTicket) Depth() int { return t.depth }

// Dependencies returns a defensive copy of dependency IDs.
func (t GeneratedTicket) Dependencies() []string {
	out := make([]string, len(t.dependencies))
	copy(out, t.dependencies)
	return out
}

// DependencyOrder is a topologically sorted ticket execution order.
type DependencyOrder struct {
	orderedIDs []string
}

// NewDependencyOrder creates a DependencyOrder value object.
func NewDependencyOrder(orderedIDs []string) DependencyOrder {
	ids := make([]string, len(orderedIDs))
	copy(ids, orderedIDs)
	return DependencyOrder{orderedIDs: ids}
}

// OrderedIDs returns a defensive copy of ordered ticket IDs.
func (d DependencyOrder) OrderedIDs() []string {
	out := make([]string, len(d.orderedIDs))
	copy(out, d.orderedIDs)
	return out
}
